tools/pbprint.cc

```diff
@@ -3,110 +3,131 @@
 #include <iomanip>
 #include <iostream>
 #include <vector>
-#include <asio.hpp>
-#include <asio/serial_port.hpp>
-#include <asio/signal_set.hpp>
 #include <google/protobuf/io/coded_stream.h>
-#include "packet/frame.h"
+#include "cobs.h"
 #include "pose.pb.h"
 #include "simulation.pb.h"
+#include <asio.hpp>
+#include <asio/serial_port.hpp>
+#include <asio/signal_set.hpp>
 
 namespace {
+
+    using std::size_t;
+
     asio::io_service io_service;
     asio::serial_port port(io_service);
 
-    constexpr size_t buffer_size = 1024;
-    uint8_t rx_buffer[buffer_size];
-    uint8_t sim_buffer[buffer_size];
-    uint8_t frame_buffer[buffer_size];
-    size_t frame_buffer_index = 0;
-
-    SimulationMessage msg;
+    // serial port
+    //   --[read]--> serial_buffer
+    //   --[cobs decode]--> packet_buffer
+    //   --[protobuf deserialize]--> message_object
+    constexpr size_t SERIAL_BUFFER_CAPACITY = 2000;
+    uint8_t serial_buffer_start[SERIAL_BUFFER_CAPACITY];
+    uint8_t * serial_buffer_write = serial_buffer_start;
+    uint8_t * const serial_buffer_end = serial_buffer_start + SERIAL_BUFFER_CAPACITY;
 
     void handle_read(const asio::error_code&, size_t);
 
     void start_read() {
-        port.async_read_some(asio::buffer(rx_buffer, buffer_size),
-                std::bind(&handle_read,
-                    std::placeholders::_1,
-                    std::placeholders::_2));
+        const size_t serial_buffer_remaining = serial_buffer_end - serial_buffer_write;
+        port.async_read_some(
+            asio::buffer(serial_buffer_write, serial_buffer_remaining),
+            &handle_read
+        );
     }
 
-    void handle_read(const asio::error_code& error, size_t bytes_transferred) {
+    void deserialize_packet(const uint8_t * const packet_buffer_start, const size_t packet_buffer_length) {
+        google::protobuf::io::CodedInputStream input(
+            packet_buffer_start,
+            packet_buffer_length
+        );
+
+        uint32_t size;
+        if (!input.ReadVarint32(&size)) {
+            std::cerr << "Unable to decode packet size from serialized data." << std::endl;
+            exit(-1);
+        }
+
+        SimulationMessage msg;
+        if (!msg.ParsePartialFromCodedStream(&input)) {
+            std::cerr << "Unable to parse protobuf message." << std::endl;
+            exit(-1);
+        }
+
+        if (!input.ConsumedEntireMessage()) {
+            std::cerr << "Entire message not consumed. Number of extra bytes: " << input.BytesUntilLimit() << std::endl;
```

Comment:
That is true. However the condition should never evaluate to true so it will only make the code more complex while not adding anything.